### PR TITLE
fix wrap intrinsic height calculation

### LINKED ISSUE #61204: Wrap widget doesn't layout correctly in certain cases, like Column does
When using the technique listed here, under the heading "Expanding content to fit the viewport": https://api.flutter.dev/flutter/widgets/SingleChildScrollView-class.html,

a Column widget works very well, but a Wrap does not. Is there something maybe missing from the way the Wrap widget lays itself out that could be improved to act more like a Column?

Essentially, the following sample shows the problem. If you run this app on a device, it should start out with a view that can be scrolled so all children are accessible. Hitting the button at the top to switch to using a Wrap will cause an Overflow.
Switching back to a Column, you can scroll again, and if you change the number in the text field to a smaller number, you can get the Wrap widget to work well as long as no Overflow happens.

```dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: WrapHomePage(),
    );
  }
}

class WrapHomePage extends StatefulWidget {
  const WrapHomePage();

  @override
  _WrapHomePageState createState()  => _WrapHomePageState();
}

class _WrapHomePageState extends State<WrapHomePage> {

  bool _useColumn = true;
  int _numTexts = 8;

  List<Widget> _buildTexts(int numTexts) {
    List<Widget> texts = [];
    for (int i=0; i<numTexts; i++) {
      texts.add(Padding(
        padding: const EdgeInsets.only(top: 24.0),
        child: Text(
            "Hello, this is some long text. It can go on and on. This test is to make sure we can scroll"),
      ));
    }
    return texts;
  }

  @override
  Widget build(BuildContext context) {

    final texts = _buildTexts(_numTexts);

    return Scaffold(
        body: SafeArea(
          child: Padding(
            padding: const EdgeInsets.all(24),
            child: LayoutBuilder(
              builder: (context, constraints) {
                return SingleChildScrollView(
                    child: ConstrainedBox(
                      constraints: BoxConstraints(
                        minHeight: constraints.maxHeight,
                      ),
                      child: IntrinsicHeight(
                        child: DefaultTextStyle(
                          style: TextStyle(fontSize: 20, color: Colors.black),
                          child: Column(
                            mainAxisAlignment: MainAxisAlignment.start,
                            mainAxisSize: MainAxisSize.max,
                            children: [
                              FlatButton(
                                  onPressed: () {
                                    setState(() {
                                      _useColumn = !_useColumn;
                                    });
                                  },
                                  child: Text(_useColumn ? "Switch to Wrap!" : "Switch to Column!", style: TextStyle(fontSize: 20, color: Colors.black))
                              ),
                              if (_useColumn)
                                Column(
                                  key: ValueKey("header"),
                                  children: texts,
                                )
                              else
                                Wrap(
                                  key: ValueKey("header"),
                                  children: texts,
                                ),
                              Expanded(
                                  key: ValueKey("middle"),
                                  child: Container(
                                    color: Colors.red,
                                    child: Center(
                                        child: TextField(
                                          onChanged: (value) {
                                            setState(() {
                                              _numTexts = int.parse(value);
                                            });
                                          },
                                        )
                                    ),
                                  )
                              ),
                              Padding(
                                key: ValueKey("footer"),
                                padding: const EdgeInsets.only(top: 24.0),
                                child: Column(
                                  children: [
                                    Text("Final message")
                                  ],
                                ),
                              )
                            ],
                          ),
                        ),
                      ),
                    )
                );
              }
            ),
          )
        )
    );
  }
}
```

### LINKED ISSUE #63302: IntrinsicHeight around Wrap(Text) is incorrect
b/163101099

Intrinsic Height calculation for Wrap() widgets with overflowing text is incorrect, causing clipping/renderflex overflows.

This occurs in our app when we have maximum accessibility text size.

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  State createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  var useIntrinsicHeight = false;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'IntrinsicHeight',
      theme: ThemeData(primarySwatch: Colors.blue),
      darkTheme: ThemeData(primarySwatch: Colors.blue),
      home: MediaQuery(
        data: MediaQueryData(textScaleFactor: 3.2),
        child: Scaffold(
          appBar: AppBar(title: Text('IntrinsicHeight Test')),
          body: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              _wrapWidget(
                Wrap(
                  children: [
                    Text('Longer text 12345678'),
                    Text('Longer text 12345678'),
                  ],
                ),
              ),
              Center(
                child: RaisedButton(
                  onPressed: () =>
                      setState(() => useIntrinsicHeight = !useIntrinsicHeight),
                  child: Text('Toggle wrapping'),
                ),
              )
            ],
          ),
        ),
      ),
    );
  }

  Widget _wrapWidget(Widget child) {
    if (!useIntrinsicHeight) {
      return Container(
        color: Colors.blue,
        child: child,
      );
    }

    return Container(
      color: Colors.green,
      child: IntrinsicHeight(child: child),
    );
  }
}
```

The expectation is that there should be no difference in height when IntrinsicHeight is toggled on or off.

### PR DESCRIPTION
**Pull request by chunhtai**
## Description

Possible fix to wrap intrinsic problem

## Related Issues

Fixes https://github.com/flutter/flutter/issues/61204
Fixes https://github.com/flutter/flutter/issues/63302

## Tests

I added the following tests:

TBD

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by Piinks — COMMENTED** [review]
This proposal looks good to me! Tests all check out too!

File: packages/flutter/test/rendering/wrap_test.dart [hunk 1]
```diff
@@ -70,6 +70,29 @@ void main() {
...
+
+    final RenderWrap renderWrap = RenderWrap();
+
+    children.forEach(renderWrap.add);
+
+    renderWrap.spacing = 0;
+    renderWrap.runSpacing = 0;
+    renderWrap.direction = Axis.horizontal;
+
+    const double oneLine = 14.0;
```
Comment (goderbauer) [inline packages/flutter/test/rendering/wrap_test.dart:89]: Maybe comment where this is coming from or force the paragraph to actually use this size.

File: packages/flutter/test/rendering/wrap_test.dart [hunk 2]
```diff
@@ -70,6 +70,29 @@ void main() {
...
+    final List<RenderBox> children = <RenderBox>[
+      RenderParagraph(
+        const TextSpan(text: 'A very very very very very very very very long text'),
+        textDirection: TextDirection.ltr,
+      ),
+    ];
+
+    final RenderWrap renderWrap = RenderWrap();
+
+    children.forEach(renderWrap.add);
```
Comment (goderbauer) [inline packages/flutter/test/rendering/wrap_test.dart:83]: Why not just renderWrap.add(RenderParagraph(...))?

**Comment by chunhtai** [comment]
ready for another look @goderbauer @Piinks

File: packages/flutter/test/rendering/wrap_test.dart [hunk 3]
```diff
@@ -70,6 +70,54 @@ void main() {
     expect(renderWrap.computeMinIntrinsicHeight(79), 250);
   });
 
+  test('Compute intrinsic height test for width-in-height-out children', () {
+    const double lineSize = 15.0;
```
Comment (goderbauer) [inline packages/flutter/test/rendering/wrap_test.dart:74]: nit: lineHeight?

File: packages/flutter/test/rendering/wrap_test.dart [hunk 4]
```diff
@@ -70,6 +70,54 @@ void main() {
...
+    renderWrap.runSpacing = 0;
+    renderWrap.direction = Axis.horizontal;
+
+    expect(renderWrap.computeMaxIntrinsicHeight(double.infinity), lineSize);
+    expect(renderWrap.computeMaxIntrinsicHeight(600), 2 * lineSize);
+    expect(renderWrap.computeMaxIntrinsicHeight(300), 3 * lineSize);
+  });
+
+  test('Compute intrinsic width test for height-in-width-out children', () {
+    const double lineSize = 15.0;
```
Comment (goderbauer) [inline packages/flutter/test/rendering/wrap_test.dart:96]: same.

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**